### ISSUE TITLE
"left side" in error message

### ISSUE THREAD

**Issue by shriram** [brownplt/pyret-lang#868]
A student came in with this error today.

![2016-09-12 13 02 21](https://cloud.githubusercontent.com/assets/75030/18456747/6ca3692c-7921-11e6-9369-05b6f6cae2fe.jpg)

I confess _I_ was not quite sure what the error message was saying. By ignoring the error message, I was able to focus him in on the contract for `overlay` and get him unstuck. But nothing about this error message struck me as particularly intuitive.

**Comment by blerner** [comment]
IIRC, this was your suggested wording for avoiding talking about the "function being applied" or the "applicant" or whatever.  The fact that "left" and "right" are not the right directions when you've aligned your code vertically is...frustrating, but not something we can realistically address.

**Comment by jswrenn** [comment]
@shriram: I think @blerner touches on one of the issues here. Another issue, as you say, is that the contract _does_ matter. This error message is missing something that arity mismatches involving user-defined functions have: the function header is embedded in the error message:
![screenshot-code pyret org 2017-06-19 13-06-55](https://user-images.githubusercontent.com/3820879/27296810-5bfffd0a-54f0-11e7-9e7b-bbfe3bc1ae3e.png)

IIRC, there was some objection to doing this for library code. 

@blerner: it's not just vertically organized code where `left` and `right` seem odd. You _can_ (but probably won't) construct programs where `left` and `right` are _completely_ misleading:
```
    true
  + 
2
```

However, I think that formatting is probably sufficiently weird that we don't need to go through hoops to accomodate it.

**Comment by blerner** [comment]
Yes: the objection was that you wanted to require that we ship the source code of Pyret as part of CPO, and re-parse it at runtime in order to generate your snippets! And that's not even technically possible for functions defined by JavaScript libraries instead of Pyret ones...

**Comment by jswrenn** [comment]
> Yes: the objection was that you wanted to require that we ship the source code of Pyret as part of CPO, and re-parse it at runtime in order to generate your snippets!

The objection shouldn't be that I want something, it should be that it's a bad idea. If including snippets of definitions provides useful context, then _not_ including it should come with a compelling argument that the costs outweigh the benefits. 

> And that's not even technically possible for functions defined by JavaScript libraries instead of Pyret ones

You are right in that the same machinery couldn't be used, but including a function signature for JavaScript libraries is almost easier, since many of our JavaScript libraries include a structured representation of their provided functions' type signatures.

**Comment by blerner** [comment]
Jack, we've hashed this out several times.  It's not "that you wanted it" that's the problem; it's that it's a technically bad idea to _require_ that the _only_ way to get a fully-working CPO experience for any Pyret program, Pyret itself included, is to ship the source of said program.  That precludes separately-compiled libraries (like teacher libraries), or black-box libraries where we don't want to show students the code.

**Comment by jpolitz** [comment]
@blerner There's an argument for showing function headers without showing full code, though, right?  And the idea that we could _synthesize_ the function header to show based on static information.  That is doable without having all the code.  So I can see the value of coming up with a solution that handles this.

@jswrenn the other argument here is that the infrastructure to show function headers _does not exist_ and would need to be built.  It's clearly a technical challenge to pull that off well, especially linking the required source tracking for higher-order functions and methods with the source and/or synthesized headers.  We're moving towards it incrementally with source maps, so hopefully we will get there.  In the meantime, it's not an OK argument to say “the error messages just have internal errors.”

**Comment by jswrenn** [comment]
> it's a technically bad idea to require that the only way to get a fully-working CPO experience for any Pyret program, Pyret itself included, is to ship the source of said program

This is _already_ the case; compare [Shriram's screenshot](https://github.com/brownplt/pyret-lang/issues/868#issue-176513078) to [my screenshot](https://github.com/brownplt/pyret-lang/issues/868#issuecomment-309506557). I don't know what the takeaway from your comment is: should we remove these snippets from messages that already have them? This seems like a Vonnegutian solution.

**Comment by blerner** [comment]
@jswrenn yes, if user code exists, we can use it. Your claim was about "there was some objection to doing this for library code"; I'm reminding this discussion what that objection was.

@jpolitz, yes that would be cool, and we almost have that infrastructure available with the static module information we now generate. We don't preserve parameter names (yet) or the whitespace or formatting (likely never), so there's still a UI problem of exactly what to display if we tried to synthesize such a signature.

**Comment by shriram** [comment]
Why can't the compiled code of a closure include the bits of source information needed to reconstruct this kind of error message? It would lead to a bit of code bloat (some of which could be further reduced if, say, we store only names but not the full text of the header), but it would not require the entire source code to have to ship, and would enable separate compilation to also work. It would make the generated code a bit larger but save the price of having to ship the entire source, which would be much worse.

**Comment by blerner** [comment]
On a technical level, adding stuff to a closure is a non-starter of an engineering approach.  On a conceptual level, yes, it would be nice to have more static information.

On the one hand: Right now, Pyret generates bajillions of closures.  Adding _more_ information to each of them will incur *major* memory bloat at runtime.  This is the same reason we don't currently add the docstrings to our closures: it's too darned expensive.  (And yes, signatures are comparably memory-expensive to docstrings: strings are a sequence of bytes and most are relatively short; signatures will need to be a few nested objects and strings and such, so I think it'll roughly balance out.)

On the other hand: docstrings and signatures are static information.  They should be associated with _functions_, not _closures_.  We have the beginnings of static information about our modules already -- that's how we're providing datatypes across modules, after all -- and conceptually I'm all in favor of enhancing those files to basically be .pdb or symbol files for Pyret programs, _provided_ the semantics of Pyret/CPO/whatever IDE is not defined in terms of them.  We have to be able to provide the equivalent of a "stripped executable", a lean, minimal .jarr file that can run as fast as possible but might give up clarity of error messages, and we have to be able to do that at a per-module level, to support teachers who want to provide black-box modules to their students.

On the gripping hand: it's not trivial even to define what you want, here.  
- If we only store names but not the full text of the header, then you're tacitly encouraging one potential source of confusion, that the error-message-rendered signatures do not match how they were typeset in the code.  (This was one of the major sources of frustration with ML's type errors that we ran into with Seminal.)  And I don't think we want the engineering burden of maintaining two code paths "if we have the raw source, and if it's parseable as a fragment, then use the existing code, otherwise synthesize and pretty-print the signature from the stored metadata."  The error-rendering code is already heinously long and tricky to maintain, and the "same" message gets rendered at least three separate times, for three separate fallback conditions; I don't think we want to add a fourth.
- Not all functions even have names for their arguments: `_.x`, or any other underscore-function, for example.

I want to try to engineer our compiled function representations to be even less heavyweight than they already are (ideally: I want them to just be JS functions), and I do not want to consider any representation choice that makes that even harder to achieve.

**Comment by jpolitz** [comment]
It's totally doable to include source information for every function in a purely technical sense.  It involves numerous decisions about the various kinds of headers we have (Ben points out several kinds of non-standard-looking functions), and reasonable design decisions to make sure it stays static and not take up lots of space on dynamic values.  So it's _not_ something where we just flip a switch and add a line of code and suddenly it's there.

The comment that sparked this discussion—that there was an objection to doing this for library code—is not a philosophical one.  It's simply that the infrastructure for doing this needs to be designed and built, and there had better be a reasonable compromise for rendering errors in the meantime, along with some kind of plan or start of a design for including the source.

Again, sourcemaps move towards this, but aren't a complete solution.  There will be several pieces to a good solution here.

**Comment by jswrenn** [comment]
> save the price of having to ship the entire source, which would be much worse

The costs associated with this sort of thing aren't obvious, too. If we encode this sort of information into the compiled javascript itself, then we are incurring a major size penalty _regardless_ of whether or not a program actually causes an error.

In contrast, "shipping the entire source" really just means performing a GET or local storage query for _just_ the relevant file, and doing so _only_ when an error is encountered. And, for some sense of scale: fetching `lists.arr` costs 6.3KB; fetching `pyret-spin.gif` costs 184KB.

The infrastructure for this already exists but is disabled. What we do _not_ have (and is relevant for the error message @shriram screenshotted) is any infrastructure for is providing similar information for the particular case Pyret libraries that are implemented in hand-written javascript. Getting this would be a technical challenge since there _isn't_ Pyret source code.

**Comment by jpolitz** [comment]
The infrastructure is disabled because it was implemented first by making `GET` requests to Github, then by requiring hosting the source separately and making `GET` requests in each error message that needed the snippet.  It's also not a generic solution that uses some well-known interface for fetching code for an arbitrary module, it's an ad hoc solution that relies on knowing the layout of the Pyret source repository from error-handling code.

I don't think fetching the code is the problem; due to code blowup in compilation, it's a small fraction of what we serve.

**Comment by blerner** [comment]
@jswrenn you've made that byte-size argument before; that's still not what we're arguing over (Shriram's aside to the contrary).  The infrastructure for this is brittle, ad-hoc, and exposes the directory structure of source code in a way that should be utterly irrelevant to compiled, standalone programs.

@jpolitz I don't think hosted code size is the problem at all.  I think the essential problem is defining CPO's runtime semantics in terms of the availability of source code at all.  It implies that teachers can't give black-box solutions or libraries (because the source must be made available), that Shriram can't demo the CPO-works-offline ability (because it no longer would), etc.  I really don't think we should compromise the ability of teachers to control the availability of their sources.


Look.  The point of this bug originally was that the "left side" wording was unclear.  This bug has been hijacked into talking about distributing the source code of all modules so that we can quote it in error messages, which is a non-trivial engineering task to do correctly, and not something that's going to happen in the near term.  (And we're not going to reenable the existing ad-hoc approach.)  So let's focus on figuring out whether the wording of this error message, *_in the absence of source code availability_*, can be improved.

**Comment by schanzer** [comment]
Since "left side" has been replaced by "operator", is this issue still relevant?

**Comment by jswrenn** [comment]
Unfortunately, this issue is not resolved. The text happens to be "operator" at the moment, but the last word from @shriram was that ["compared to that, I guess I'd take 'applicant'"](https://github.com/brownplt/pyret-lang/commit/bbdc98f415c27d39ca1f838b17314d0f80c4672b#commitcomment-23331726).
